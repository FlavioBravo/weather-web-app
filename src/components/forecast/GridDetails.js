import React from 'react';
import PropTypes from 'prop-types';

export const GridDetails = ({ hourlyItem }) => {
    return (
        <div>
            <br />
            <hr />
            <br />
            <h3>Hourly Forecast</h3>
            <br />
            <div className="gridDetails_container">
            {
            hourlyItem.map( hour  => (
                <p className="gridDetails_linespace" key={hour.dt} >Temperature: {hour.temp} ° | Human Perception: {hour.feels_like} ° | Humidity: {hour.humidity} % | Clouds: {hour.clouds} % | Pressure: {hour.pressure} hPa | Uvi: {hour.uvi} UV</p>
            ))
            }
            </div>
        </div>
    )
}

GridDetails.propTypes = {
    hourlyItem: PropTypes.array.isRequired
}