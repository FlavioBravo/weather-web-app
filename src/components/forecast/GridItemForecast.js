import React from 'react';
import PropTypes from 'prop-types';
import { getDayNameOfWeek } from '../../helpers/utils';
import { useFetchOpenWeatherHourly } from '../../hooks/useFetchOpenWeatherHourly';

export const GridItemForecast = ({ dt, temp_min, temp_max, weather, icon, dt_txt, lat, lon, setHourForecast }) => {

    const dayName = getDayNameOfWeek(dt_txt);
    const dateFormat = dt_txt.split(' ')[0];

    const { data: hourly, loading } = useFetchOpenWeatherHourly(lat, lon);

    const handleForecast = () => {
        setHourForecast(hourly);
    };

    return (
        <div className="animate__animated animate__fadeIn forecast_grid_item" onClick={ handleForecast } >
            <p> { dayName } </p>
            <p> { weather } </p>
            <img src={ icon } alt={ weather } height="80" width="80" />
            <div className="forecast_grid_item_footer">
                <span>{ temp_min }°</span>
                <span>{ temp_max }°</span>
            </div>
            <p> { dateFormat } </p>
        </div>
    )
}

GridItemForecast.propTypes = {
    temp_min: PropTypes.number.isRequired,
    temp_max: PropTypes.number.isRequired,
    weather: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
    dt_txt: PropTypes.string.isRequired,
}