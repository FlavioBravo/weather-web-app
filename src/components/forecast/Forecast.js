import { React, useState } from 'react';
import { useParams } from "react-router-dom";

import { useFetchOpenWeather } from '../../hooks/useFetchOpenWeather';
import { GridDetails } from './GridDetails';
import { GridItemForecast } from './GridItemForecast';

export const Forecast = () => {

    const { cityId } = useParams();

    const { data: daysForecast, loading } = useFetchOpenWeather(cityId);

    const [state, setState] = useState({
        hourlyItem: [],
        show: false
    });
    
    const setHourForecast = (hourly) => {
        if (hourly) {
            setState({
                hourlyItem: [ ...hourly ],
                show: true
            });
        } else {
            setState({
                hourlyItem: [],
                show: false
            });
        }
    };

    return (
        <>
            <h2>Forecast</h2>
            <br />
            <hr />
            <br />
            { loading && <p className="animate__animated animate__flash">Loading</p> }
            <div className="forecast_container">
              {
                  daysForecast.map( day  => (
                    <GridItemForecast
                        key={day.dt_txt}
                        {...day}
                        setHourForecast = { setHourForecast }
                    />
                  ))
              }
          </div>
            {
               state.show && 
               <GridDetails
                    hourlyItem = { state.hourlyItem }
                />
            }

        </>
    )
}
