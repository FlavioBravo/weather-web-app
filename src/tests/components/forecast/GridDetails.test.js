import React from 'react';
import { shallow } from 'enzyme';
import { GridDetails } from '../../../components/forecast/GridDetails';


describe('Pruebas en <GridDetails />', () => {

    const hourlyItem = [
            {
                dt: 1622224800,
                temp: 299.07,
                feels_like: 299.65,
                pressure: 1011,
                humidity: 74,
                dew_point: 294.08,
                uvi: 10.54,
                clouds: 42,
                visibility: 10000,
                wind_speed: 4.74,
                wind_deg: 302,
                wind_gust: 3.25,
                weather: [
                    {
                        id: 802,
                        main: "Clouds",
                        description: "scattered clouds",
                        icon: "03d"
                    }
                ],
                pop: 0
            },
            {
                dt: 1622228400,
                temp: 298.91,
                feels_like: 299.5,
                pressure: 1011,
                humidity: 75,
                dew_point: 294.14,
                uvi: 8.82,
                clouds: 45,
                visibility: 10000,
                wind_speed: 5.29,
                wind_deg: 286,
                wind_gust: 4.27,
                weather: [
                    {
                        id: 802,
                        main: "Clouds",
                        description: "scattered clouds",
                        icon: "03d"
                    }
                ],
                pop: 0.02
            },
            {
                dt: 1622232000,
                temp: 298.65,
                feels_like: 299.24,
                pressure: 1011,
                humidity: 76,
                dew_point: 294.11,
                uvi: 5.79,
                clouds: 51,
                visibility: 10000,
                wind_speed: 4.67,
                wind_deg: 273,
                wind_gust: 3.79,
                weather: [
                    {
                        id: 803,
                        main: "Clouds",
                        description: "broken clouds",
                        icon: "04d"
                    }
                ],
                pop: 0.02
            }
    ];
    const wrapper = shallow( <GridDetails hourlyItem = { hourlyItem } /> );

    
    test('debe de mostrar el componente correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    })

    test('debe de tener 6 columnas', () => {
        
        const p = wrapper.find('p');
        expect( p.length ).toBe(3);

    })

    test('debe de tener titulo', () => {
        
        const h3 = wrapper.find('h3');
        expect( h3.text().trim() ).toBe( 'Hourly Forecast' );

    })

    test('debe de tener la clase animate__fadeIn', () => {
        
        const div = wrapper.find('div.gridDetails_container');
        expect( div.length ).toBe( 1 );

    })

    test('debe de tener el contenido igual a los props', () => {
        
        const p = wrapper.find('p').first();
        expect( p.text().trim() ).toBe( 'Temperature: 299.07 ° | Human Perception: 299.65 ° | Humidity: 74 % | Clouds: 42 % | Pressure: 1011 hPa | Uvi: 10.54 UV' );

    });

})