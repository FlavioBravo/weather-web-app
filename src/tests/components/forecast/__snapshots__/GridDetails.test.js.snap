// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GridDetails /> debe de mostrar el componente correctamente 1`] = `
<div>
  <br />
  <hr />
  <br />
  <h3>
    Hourly Forecast
  </h3>
  <br />
  <div
    className="gridDetails_container"
  >
    <p
      className="gridDetails_linespace"
      key="1622224800"
    >
      Temperature: 
      299.07
       ° | Human Perception: 
      299.65
       ° | Humidity: 
      74
       % | Clouds: 
      42
       % | Pressure: 
      1011
       hPa | Uvi: 
      10.54
       UV
    </p>
    <p
      className="gridDetails_linespace"
      key="1622228400"
    >
      Temperature: 
      298.91
       ° | Human Perception: 
      299.5
       ° | Humidity: 
      75
       % | Clouds: 
      45
       % | Pressure: 
      1011
       hPa | Uvi: 
      8.82
       UV
    </p>
    <p
      className="gridDetails_linespace"
      key="1622232000"
    >
      Temperature: 
      298.65
       ° | Human Perception: 
      299.24
       ° | Humidity: 
      76
       % | Clouds: 
      51
       % | Pressure: 
      1011
       hPa | Uvi: 
      5.79
       UV
    </p>
  </div>
</div>
`;
